### PR TITLE
[extractor/dropout] Dropout season pagination

### PR DESCRIPTION
**Pull request by OverlordQ**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Handles pagination for seasons with more than 24 episodes. Also fixes test cases with accurate episode counts for seasons. 

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [X] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details><details><summary>Copilot Summary</summary>  

### Summary
📄🔗🧪

Improved the `DropoutSeasonIE` extractor to handle multiple pages of episodes. Updated the tests and the `yt_dlp/extractor/dropout.py` file accordingly.

> _We're scraping the web for some videos to see_
> _We use `DropoutSeasonIE`_
> _But the pages are many and we need to fetch more_
> _So we add pagination with `get_element_by_attribute`_

### Walkthrough
* Import `get_element_by_attribute` function from `utils.py` to use for pagination ([link](https://github.com/yt-dlp/yt-dlp/pull/7304/files?diff=unified&w=0#diff-dbeadf033fee19b4cabd2dada23bc66943284f1b67ad8fb41c177c5cd7300467R6))
* Update `playlist_count` for `DropoutSeasonIE` extractor test cases to reflect current number of episodes for "Dimension 20: Fantasy High" ([link](https://github.com/yt-dlp/yt-dlp/pull/7304/files?diff=unified&w=0#diff-dbeadf033fee19b4cabd2dada23bc66943284f1b67ad8fb41c177c5cd7300467L170-R171), [link](https://github.com/yt-dlp/yt-dlp/pull/7304/files?diff=unified&w=0#diff-dbeadf033fee19b4cabd2dada23bc66943284f1b67ad8fb41c177c5cd7300467L179-R180))
* Add new test case for `DropoutSeasonIE` extractor to cover multi-season series with pagination, such as "Breaking News No Laugh Newsroom" ([link](https://github.com/yt-dlp/yt-dlp/pull/7304/files?diff=unified&w=0#diff-dbeadf033fee19b4cabd2dada23bc66943284f1b67ad8fb41c177c5cd7300467R194-R202))
* Initialize `page_num` variable to 1 in `_real_extract` method of `DropoutSeasonIE` extractor to keep track of current page number ([link](https://github.com/yt-dlp/yt-dlp/pull/7304/files?diff=unified&w=0#diff-dbeadf033fee19b4cabd2dada23bc66943284f1b67ad8fb41c177c5cd7300467R210))
* Add pagination logic to `_real_extract` method of `DropoutSeasonIE` extractor using `get_element_by_attribute`, `get_elements_by_class`, and `url_result` functions to download and extract all episodes of a paginated series ([link](https://github.com/yt-dlp/yt-dlp/pull/7304/files?diff=unified&w=0#diff-dbeadf033fee19b4cabd2dada23bc66943284f1b67ad8fb41c177c5cd7300467R220-R234))

</details>

### CODE REVIEW

**Comment by pukkandan** [comment]
Try to return a PagedList instead, or atleast a generator.

**Comment by OverlordQ** [comment]
Second pass. The `try/except` seems hacky but only way I saw to not barf when `len(page 1) == page_max`

File: yt_dlp/extractor/dropout.py [hunk 1]
```diff
@@ -190,29 +194,44 @@ class DropoutSeasonIE(InfoExtractor):
...
+        entries = OnDemandPagedList(functools.partial(
+            self._fetch_page, url, season_id), self._PAGE_SIZE)
 
-        return {
-            '_type': 'playlist',
-            'id': join_nonempty(season_id, current_season.lower().replace(' ', '-')),
-            'title': join_nonempty(season_title, current_season, delim=' - '),
-            'entries': entries
-        }
+        return self.playlist_result(entries, playlist_id=f'{season_id}-season-{season_num}', playlist_title=f'{season_title} - Season {season_num}')
```
Comment (bashonly) [inline yt_dlp/extractor/dropout.py:237]: line too long
```suggestion
        return self.playlist_result(
            entries, f'{season_id}-season-{season_num}', f'{season_title} - Season {season_num}')
```

&nbsp;
or better imo, you could inline entries (and remove var above)
```suggestion
        return self.playlist_result(
            OnDemandPagedList(functools.partial(self._fetch_page, url, season_id), self._PAGE_SIZE),
            f'{season_id}-season-{season_num}', f'{season_title} - Season {season_num}')
```

**Comment by OverlordQ** [comment]
Changes made, but still the nested exception issue.

File: yt_dlp/extractor/dropout.py [hunk 2]
```diff
@@ -190,29 +195,30 @@ class DropoutSeasonIE(InfoExtractor):
...
+                'id': 'breaking-news-no-laugh-newsroom-season-3',
+                'title': 'Breaking News No Laugh Newsroom - Season 3'
+            }
         }
     ]
 
+    def _fetch_page(self, url, season_id, page):
+        page += 1
+        webpage = self._download_webpage(
+            f'{url}?page={page}', season_id, note=f'Downloading page {page}')
```
Comment (bashonly) [inline yt_dlp/extractor/dropout.py:213]: My bad, I was wrong. yt-dlp will catch the error and continue with playlist extraction, but the extractor & tests do fail. Let's do this then:
```suggestion
        webpage = self._download_webpage(
            f'{url}?page={page}', season_id, note=f'Downloading page {page}', expected_status=400)
```

**Review by bashonly — APPROVED** [review]

**Pull request merged**